src/include/nfs/toposort.ycp

```diff
@@ -0,0 +1,137 @@
+                                           ``($[vertex: 0]));
+        foreach (string vertex, list<string> targets, dag, ``{
+            foreach (string target, targets, ``{
+                in_degree[target] = in_degree[target]:0 + 1;
+            });
+        });
+
+        while (size (in_degree) > 0)
+        {
+            // get the vertices that can go next because they have zero in degree
+            map<string, integer> next_m =
+                (map<string, integer>) filter (string vertex, integer d,
+                                               in_degree,
+                                               ``(d == 0));
+            if (size (next_m) == 0)
+            {
+                // the graph is cyclic!
+                break;
+            }
+            foreach (string vertex, integer dummy, next_m, ``{
+                out = add(out, vertex);
+            });
+            // remove the vertices
+            in_degree = (map<string, integer>) filter (string vertex, integer d,
+                                                       in_degree,
+                                                       ``(d != 0));
+            // remove the edges that were leading from them
+            foreach (string vertex, integer dummy, next_m, ``{
+                foreach (string target, dag[vertex]:[], ``{
+                    in_degree[target] = in_degree[target]:0 - 1;
+                });
+            });
+        }
+
+        list<string> rest = (list<string>) maplist (string k, integer v, in_degree, ``( k )); //mapkeys
+        if (size (rest) > 0)
+        {
+            y2error ("Cyclic subgraph found, remainder has %1 nodes: %2",
+                     size (rest), rest);
+        }
+        return [out, rest];
+    }
+
+    /**
+     * TopologicalSort as implemented above only works on graphs
+     * with string vertices.
+     * Now we will
+     * 1. generalize it to arbitrary data ("things")
+     * 2. exploit the order of keys in map<string, ...> to produce a *stable*
+     *    topological sort (one that preserves the order of unrelated vertices)
+     */
+    typedef any thing_t;
+    typedef string vertex_t;
+
+    /**
+     * @param things a list of things
+     * @param edge   a function determining whether
+     *               a directed edge exists between two things
+     * @return [out, rest]<br>
+     *  out: a list of things in topological order
+     *       (starting with things that have no *incoming* edge)<br>
+     *  rest: a list, empty if the graph was acyclic, otherwise it is
+     *   a superset of the things forming the cycle
+     *   and "out" is a partial result
+     *
+     */
+    list<list<thing_t> > StableTopologicalSort(list<thing_t> things,
+                                               boolean(thing_t, thing_t) edge) {
+      // The stability of the sort is achieved
+      // by using zero-padded numeric strings as graph vertices.
+      // For 31337 things we will need 5-character vertices
```

Comment:
so can you please update sentence to have `For example 31337` as it slightly confuse me what it stand for.